### PR TITLE
Improve homomorphisms code

### PR DESCRIPTION
**Pull request by james-d-mitchell**
This PR improves the code for `DigraphEmbedding` and related, after a suggestion by @flsmith just before the holidays. 

While writing code coverage tests for `homos.c` I also discovered some bugs in the homomorphism finding code when a partial homomorphism was specified. If the values entered in the partial homomorphism were not orbit representatives of some pointwise stabiliser in the automorphism group of the range graph, then in some cases `HomomorphismDigraphsFinder` returned an empty list even when there were homomorphisms extended the partially specified homomorphism. This PR also fixes this bug. 

This breaks backwards compatibility of `HomomorphismDigraphsFinder` since the 7th argument is no longer a boolean but instead can be 0, 1, or 2.

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
On second thoughts, it wouldn't be hard to not break backwards compatibility and to still allow `true` or `false` as the 7th argument of `HomomorphismDigraphsFinder` interpreted in the same way as before.

**Comment by wilfwilson** [comment]
@james-d-mitchell Please let me know when/if you've done this change, and then I'll review.

**Comment by james-d-mitchell** [comment]
@wilfwilson and @flsmith I've more or less finished this now, I probably need to add some tests, and there are a couple of other things I'd like to try out, but if you have the time, you might want to start reviewing this now. 

This PR started by just adding some new features to the homomorphism finding code, and has ended up with me completely refactoring the code. 

On the plus side:

* there are more features now (you can specify the order of the vertices, which sometimes makes a profound impact on the run time)
* the second example in `tst/extreme/grahom.tst` takes approximately 2.2 seconds (in master), and takes approx. 120ms with this PR
* the `partial_map` argument of the function `HomomorphismDigraphsFinder` never worked properly before (it did not return non-homomorphisms, but did sometimes return that there are no homomorphisms extending the given partial one, when, in fact there were). This is fixed.
* less memory is allocated per call to `HomomorphismDigraphsFinder`, in particular, except for permutations, and the memory used by Bliss, memory is only allocated once by HomomorphismDigraphsFinder`, the first time it is called.
* the code is much shorter, there is less duplicate code, and hopefully it is now more sensibly organised.

**Comment by flsmith** [comment]
The internet at the Burn has been pretty shocking so I haven't been able to get this done. I plan to look at it this afternoon.

**Comment by james-d-mitchell** [comment]
Ok, I've finally finished playing with this. I need to add some tests for code coverage. For info, there is now no memory allocation after the first ever call anywhere in the homomorphism finding code. This has not a huge impact on single examples, but can make repeated calls to the homomorphism finding code significantly faster. 

I will also have to make the changes suggested by @wilfwilson.

File: gap/grahom.gi [hunk 1]
```diff
@@ -431,42 +386,72 @@ InstallMethod(EpimorphismsDigraphs, "for a digraph and a digraph",
...
+end);
+
+InstallMethod(EmbeddingsDigraphs, "for a digraph and a digraph",
+[IsDigraph, IsDigraph],
+function(gr1, gr2)
+  local hom, aut;
+  hom := EmbeddingsDigraphsRepresentatives(gr1, gr2);
+  aut := List(AutomorphismGroup(DigraphRemoveAllMultipleEdges(gr2)),
+              AsTransformation);
+  return Union(List(aut, x -> hom * x));
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/grahom.gi]: Good suggestion, but the others (`DigraphHomomorphisms` and chums) look like this, so I think I'll leave it for now at least.

**Comment by wilfwilson** [comment]
@flsmith are you going to look at this too?

**Comment by flsmith** [comment]
Nearly done, will be completed tomorrow morning.

File: src/homos.c [hunk 2]
```diff
@@ -1,1816 +1,1861 @@
...
-  DIGRAPHS_ASSERT(i < graph->nr_vertices && j < graph->nr_vertices);
-
-  return get_bit_array(graph->neighbours[i], j);
+static ALWAYS_INLINE uint16_t
+                     graph_mono_update_conditions(uint16_t const depth,
+                                                  uint16_t const last_defined,
+                                                  uint16_t const vertex) {
+  push_conditions(
+      CONDITIONS, depth, vertex, GRAPH2->neighbours[MAP[last_defined]]);
+  set_bit_array(get_conditions(CONDITIONS, vertex), MAP[last_
```
Comment (flsmith) [inline src/homos.c]: digraph2_obj -> digraph1_obj

Comment (flsmith) [inline src/homos.c]: and here

Comment (flsmith) [reply src/homos.c]: The preview here seems to be completely the wrong line, my comment is for line 1725

Comment (flsmith) [reply src/homos.c]: ... line 1728

File: src/conditions.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+};
+
+typedef struct conditions_struct Conditions;
+
+//! Returns a pointer to a Conditions with one complete row where every bit is
+//! set to true.
+Conditions* new_conditions(uint16_t const nr1, uint16_t const nr2);
+
+//! Returns a pointer to a Conditions with one complete row where every bit is
+//! set to true.
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply src/conditions.h]: `true` is correct in this case.

File: src/homos.c [hunk 4]
```diff
@@ -1,1816 +1,1861 @@
...
-  ORB_LOOKUP = new_bit_array(NR2);
-  REPS       = malloc(NR1 * sizeof(BitArray*));
-
-  // initialise the <map> and store the sizes in the conditions.
-  for (i = 0; i < NR1; i++) {
-    REPS[i] = new_bit_array(NR2);
-    MAP[i]  = UNDEFINED;
+// Rewrite the global variable MAP so that it contains a homomorphism of the
+// internal (di)graph, and not the possibly distinct (but isomorphic) GAP level
+// (di)graph.  This should be called before any calls to the hook functions
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply src/homos.c]: `before` is correct in this case too.

[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply src/homos.c]: No sorry I looked at the wrong place, you're right

**Comment by james-d-mitchell** [comment]
@flsmith I think I've addressed all of your issues.

**Comment by wilfwilson** [comment]
@james-d-mitchell you pushed since @flsmith's approval – are you still making changes, or have you finished?

**Comment by james-d-mitchell** [comment]
It was just a rebase, I’m completely finished now.

**Pull request merged**